README
FLOTSAM: Forward-Lobe Two-Stream Radiance Model

Contact: Robin Hogan <r.j.hogan@ecmwf.int>

FLOTSAM is a fast solar radiance model intended to be suitable for
assimilating solar radiances affected by clouds and aerosols into
models. Unlike models based on look-up tables, it can take as input an
arbitrary layering of clouds, aerosols and Rayleigh-scattering clear
sky. It is still under heavy development.

It is coded in C++ using the Adept array and automatic-differentiation
library to compute the Jacobian matrix.  The library presents a C
interface (see include/flotsam.h).  An example driver program is
bin/flotsam.cpp.  Many test scripts are in the test directory.

WARNING: THIS CODE IS UNDER DEVELOPMENT AND MANY OF THE SETTINGS ARE
SUBJECT TO CHANGE.  THIS VERSION HAS HAD ITS SETTINGS ADJUSTED TO BEST
REPRESENT AEROSOLS, ALTHOUGH IT IS KNOWN THAT THE WAY THE PHASE
FUNCTION IS PARAMETERIZED DOES NOT YET WORK WELL ENOUGH FOR SOME
AEROSOL PHASE FUNCTIONS.  THIS WILL BE IMPROVED IN THE NEXT VERSION.


INSTALLATION

Before compiling FLOTSAM you will need to install the Adept library
(version 2+) from http://www.met.reading.ac.uk/clouds/adept, ensuring
that it finds a working BLAS library.

Configure the compilation by running the ./configure script. If you
are using the GNU compiler then it is recommended that you run the
./configure_fast.sh script instead, as it chooses the fastest options
for this compiler.  If you installed Adept in a non-system directory
then you will need to tell configure where it is.  The easiest way to
do this is to call configure_fast.sh with the ADEPT_PREFIX environment
variable set:

ADEPT_PREFIX=/path/to/adept ./configure_fast.sh

This tells configure to look for Adept header files in
/path/to/adept/include and the Adept shared library in
/path/to/adept/lib64.

Then compile with "make".  You can the try the test scripts in the
test directory.

To install you should have called the configure or configure_fast.sh
scripts with --prefix=/path/to/flotsam argument to specify the
location where you want to install it - at this location you will then
find the header file include/flotsam.h, the executable bin/flotsam,
the static library lib64/libflotsam.a and the shared library
lib64/flotsam.so.0.0.0.


TESTING

The test directory contains a number of test scripts described in the
test/README file.  These run the bin/flotsam executable, which call
the algorithm using a fixed phase function at all heights.  Its usage
is:

flotsam [param1=value1] [param2=value2...] file1.cfg [file2.cfg...]

where at least one ascii configuration file (*.cfg) must be provided;
these files contain lists of configuration options and there are a
number of examples in the test directory.  Configuration options can
be overridden on the command line.  The scripts call this executable.
Its output (on standard output) is of the following format:

Line 1: 5 numbers:
1: Number of solar zenith angles simulated
2: Number of instrument zenith angles simulated
3: Number of azimuth angles simulated
4: Number of optical depths simulated
5: "N" = number of values reported for each geometry and optical depth

Subsequent lines: 4+N numbers:
1: Solar zenith angle (degrees)
2: Instrument zenith angle (degrees)
3: Difference in azimuth angle between sun and instrument (degrees)
4: Optical depth of particulates
5: Normalized radiance
6: (if present) Direct contribution to normalized radiance
7: (if present) Lobe contribution to normalized radiance
8: (if present) Diffuse contribution to normalized radiance
9: (if present) Surface contribution to normalized radiance
10:(if present) Two-stream estimate of TOA upwelling normalized flux

The program also reports its settings and timings on standard error.
The algorithm is designed to be run many times with the same geometry
(as would be expected in an iterative retrieval algorithm or data
assimilation system).  Therefore, the timings reported are split into
setup calculations and reflectance calculations (there are usually
many more of the latter).

ChangeLog
version 0.5.24
	- Use the gettimeofday() rather than clock() timing mechanism by
	  default, since the latter seems to be unreliable on some systems
	- Added bin/ocean executable to output Cox-Munk albedo

version 0.5.23 (11 March 2022)
	- C/Fortran interface to flotsam_analyse_phase_functions, which
	  now allows for phase functions to be provided on an irregular grid
	  in angle, useful for strongly forward-peaked phase functions
	- Halved resolution of ocean BRDF look-up table for speed

version 0.5.22 (6 March 2022)
	- Fixed Fortran interface (tested using a the flotsam branch of ecRad)
	- In phase_function_component.cpp, convolve phase_function now
	  correctly performed up to pf(FLOTSAM_NUM_SMOOTHING_VALUES,__),
	  thanks to Jeronimo Escribano

version 0.5.20 (16 February 2018)
	- Added SZA-dependent scaling of lobe radiances at low optical
	  depths to account for a similar effect that has already been
	  accounted for in diffuse fluxes
	- Tuned settings a little for aerosol cases

version 0.5.19 (16 February 2018)
	- Reverted prefactor on lobe calculation ("revision2") to ensure
	  energy conservation in fluxes
	- Reverted secants used in beam-to-diffuse exchange ("revision5")
	  to improve reflectance at low viewing angle
	- Generalized two-stream gamma1/gamma2 calculation since PIFM is
	  not theoretically correct for near isotropic phase functions
	- Increased scaling-up of diffuse fluxes since reference optical
	  depth should probably be only half the optical depth of the
	  profile
	- Tunables controlled more from base.hpp

version 0.5.18 (13 Febrary 2018)
	- Fix calculation of prob_direct_to.diffuse_[up|dn]
	- Fixed bug in polynomial smoothed phase function lookup
	- Instead of reading in phase_function_file, can specify the 6
	  components with phase_function_components, which uses the fitted
	  phase function for the direct component

version 0.5.17 (29 January 2018)
	- Predict the angular variance of the phase function and use to
	  interpolate smoothed phase function
	- Added mode=internals which outputs Matlab script of internal
	  arrays, and test/do_internals.sh for performing such runs

version 0.5.16 (8 January 2018)
	- <complex.h> -> <complex> in bi_directional_surface_reflectivity.hpp
	- Removed unneeded "typename" from reflectance.cpp
	- New set of 6 basis functions for fitting phase functions
	  (replacing 3)

version 0.5.15 (11 December 2017)
	- Formula to scale up radiances at low optical depth now applied
	  to diffuse only, not lobe

version 0.5.14 (20 November 2017)
	- Formula to scale up radiances at low optical depth now a fit to
	  theory
	- Left in optimum configuration for aerosol

version 0.5.13 (11 November 2017)
	- Solved bug in Meador & Weaver to use sqrt, and separate path in
	  case of conservative scattering
	- Added option to scale up diffuse contribution to radiances at
	  very low optical depth, as found that in Rayleigh atmospheres,
	  flux from integrating over radiances was underestimated compared
	  to upwelling two-stream flux
	- do_analyse_phase_functions.sh now analyses aerosol phase
	  functions as well as cloud
	- Left in new optimum configuration for aerosol (see base.hpp)
	- Added a basic README file

version 0.5.12 (2 November 2017)
	- Clarified optimum settings for aerosol and cloud in base.hpp
	- Added slope_dist_shape option to Cox&Munk model
	- Identified but did not yet solve horrible bug in meador_weaver:
	  no sqrt on k
	- Added do_rayleigh_benchmark.sh to check fluxes for Rayleigh
	  atmosphere (identifying but not solving an erroneous resolution
	  dependence)

version 0.5.11 (22 September 2017)
	- A fraction 0.2 of the lobe part of the phase function now goes
	  into the forward hemisphere and hence into the diffuse up/dn
	  streams. This reduces the overestimate of the pseudo-specular
	  reflection at high SZA.
	- Bug-fix in diffraction scaling when ssa < 1 (previously scaled
	  ssa could go negative)
	- Diffuse scaling factor changed from up=1.1,dn=1.0 to
	  up=1.05,dn=1.05; this partially accounts for the two-stream method
	  underestimating TOA flux in pure Rayleigh conditions (base.hpp)
	- Some algorithm revisions based on the paper are turned on with
	  FLOTSAM_REVISION=1.  One is not because it degrades the
	  performance. This is controlled by FLOTSAM_REVISION2
	  (calc_beam_fluxes.hpp)
	- Components used to fit phase function now computed dynamically
	  in phase_function_component.cpp
	- Added query function to return total atmospheric Rayleigh or gas
	  absorption optical depth

version 0.5.10 (19 September 2017)
	- Completed CAMS benchmark setup in test directory

version 0.5.9 (14 September 2017)
	- Moved flotsam executable to bin directory and cleaned up

version 0.5.8 (14 September 2017)
	- Cached exponential calls resulting in 45% speed-up
	- Jacobian function pre-allocates memory which can halve execution
	  time
	- "test/flotsam mode=phase_function" can analyse phase functions
	  as well

version 0.5.7 (6 September 2017)
	- Added test program test/flotsam configurable using simple ascii
	  file
	- Modified algorithm in a few places to match paper (search for
	  ADEPT_REVISION); very small impact

version 0.5.6 (2 September 2017)
	- Modularized Cox & Munk model in OceanBRDF class

version 0.5.5 (1 September 2017)
	- Added Cox & Munk model for ocean reflectance

version 0.5.4 (21 February 2017)
	- analyse_phase_functions now takes the normalization as an argument
	- Optical depth per layer asymptotes to safe value
	- Work-arounds for dodgy Adept 1.9.8
	- Added compress_grid to compress clear-sky layers into one

version 0.5.3 (10 February 2017)
	- Corrected dependence on instrument zenith angle via modification
	  to calc_beam_fluxes.h
	- Correctly calculate Rayleigh scattering using Bucholtz (1995)
	- Phase functions (integrating to 4pi) now carried through
	  unscaled to beam_to_beam, where a final division by 4pi is
	  applied, rather than dividing them by 2 when first loaded followed

src/base.hpp
/// @file      base.hpp
/// @brief     Defines basic stuff 
/// @copyright 2017-2018 European Centre for Medium Range Weather Forcasts
/// @license   Apache License Version 2 (see the NOTICE.md file for details)

//#include <exception>

#ifndef M_PI
#define M_PI 3.14159265358979323846  /* pi */
#endif

/*
  Current optimal settings for CLOUD:
  DO NOT define FLOTSAM_REVISION*
  DO NOT define FLOTSAM_NO_PIFM
  DO NOT define FLOTSAM_DIFFUSE_SCALING_WITH_OD
  FLOTSAM_FORWARD_LOBE_DIFFUSE = 0.0
  FLOTSAM_DIFFUSE_UP_FACTOR = 1.1;
  FLOTSAM_DIFFUSE_DN_FACTOR = 1.0;

  Older optimal settings for AEROSOL:
  DO define FLOTSAM_REVISION but not FLOTSAM_REVISION2
  DO NOT define FLOTSAM_NO_PIFM
  DO NOT define FLOTSAM_DIFFUSE_SCALING_WITH_OD
  FLOTSAM_FORWARD_LOBE_DIFFUSE = 0.2
  FLOTSAM_DIFFUSE_UP_FACTOR = 1.1;
  FLOTSAM_DIFFUSE_DN_FACTOR = 1.0;

  Newer settings for AEROSOL:flotsam0513g
  DO define FLOTSAM_REVISION but not FLOTSAM_REVISION2
  DO NOT define FLOTSAM_NO_PIFM
  DO define FLOTSAM_DIFFUSE_SCALING_WITH_OD
  FLOTSAM_DIFFUSE_UP_FACTOR = 1.2;
  FLOTSAM_DIFFUSE_DN_FACTOR = 1.2;
  FLOTSAM_FORWARD_LOBE_DIFFUSE = 0.2;

  Settings for AEROSOL:flotsam0514h
  DO define FLOTSAM_REVISION but not FLOTSAM_REVISION2
  DO NOT define FLOTSAM_NO_PIFM
  DO define FLOTSAM_DIFFUSE_SCALING_WITH_OD (New formula)
  FLOTSAM_DIFFUSE_UP_FACTOR = 1.0;
  FLOTSAM_DIFFUSE_DN_FACTOR = 1.0;
  FLOTSAM_NONDIRECT_SCALING = 1.15; // Replaces those above
  FLOTSAM_FORWARD_LOBE_DIFFUSE = 0.2;
  FLOTSAM_LUT_OPTICAL_DEPTH_REF_FACTOR = 0.5; // Previously hard-coded
  OMG!!! There was a bug in the plotting that removed from the total
  half of the lobe reflectance... the scheme was then tuned to this!

 */



#ifndef FLOTSAM_BASE_HPP
#define FLOTSAM_BASE_HPP

#define FLOTSAM_REVISION 1
//#define FLOTSAM_REVISION2 1 // OFF fixes non-conservation in flux profile
#define FLOTSAM_REVISION3 1
#define FLOTSAM_REVISION3A 1
#define FLOTSAM_REVISION4 1
//#define FLOTSAM_REVISION5 1 // OFF fixes reflectance at low viewing angle
#define FLOTSAM_REVISION6 1

#define FLOTSAM_NO_PIFM 1
#define FLOTSAM_DIFFUSE_DIFFUSIVITY 2
#define FLOTSAM_BETA_PREFACTOR 0.375 // PIFM value
//#define FLOTSAM_BETA_PREFACTOR 0.5

// At low optical depth there is enhanced scattering from diffuse to
// other streams
#define FLOTSAM_DIFFUSE_SCALING_WITH_OD 1
// Distribution of diffuse angles in optically thin cloud depends on
// the optical depth considered; if we set this equal to the optical
// depth of the cloud we get this "reference" optical depth
//#define FLOTSAM_DIFFUSE_MULTIPLIER_REFERENCE_OD 2.0
// But since diffuse radiation typically originates from scattering in
// the middle of the cloud, the reference optical depth perhaps should
// be doubled
#define FLOTSAM_DIFFUSE_MULTIPLIER_REFERENCE_OD 5.0

// Do something similar with lobe reflectances at low optical depth
#define FLOTSAM_LOBE_SCALING_WITH_OD 1
#define FLOTSAM_LOBE_MULTIPLIER_REFERENCE_OD 4.0

//#define FLOTSAM_USE_OLD_COMPONENTS 1

#ifdef FLOTSAM_USE_OLD_COMPONENTS
// Original values
#define FLOTSAM_LOBE_HALF_WIDTH_PHASE_FUNC_DEG 35.0
#define FLOTSAM_LOBE_HALF_WIDTH_DIFFUSIVITY_DEG 25.0
#define FLOTSAM_LOBE_HALF_WIDTH_KERNEL_DEG 20.0
#else
#define FLOTSAM_LOBE_HALF_WIDTH_PHASE_FUNC_DEG 18.0
#define FLOTSAM_LOBE_HALF_WIDTH_DIFFUSIVITY_DEG 25.0
#define FLOTSAM_LOBE_HALF_WIDTH_KERNEL_DEG 20.0
//#define FLOTSAM_LOBE_HALF_WIDTH_DIFFUSIVITY_DEG 30.0
//#define FLOTSAM_LOBE_HALF_WIDTH_KERNEL_DEG 30.0
#endif

// Not needed any more
//#define FLOTSAM_DIFFUSE_SCALING_MAX 2.0
//#define FLOTSAM_DIFFUSE_SCALING_OD_SCALE 0.4

#include <flotsam.h>

#include <adept_arrays.h>

namespace flotsam {

  using namespace adept;

  /// Select an active versus static scalar depending on bool template
  /// argument
  template <bool IsActive>
  struct scalar {
    typedef adept::Real type;
  };

  template <>
  struct scalar<true> {
    typedef adept::aReal type;
  };


  /*
  class exception : public std::exception {
  public:
    exception(const std::string& message = "Error in FLOTSAM library")
      { message_ = message; }
    virtual const char* what() const throw() { return message_.c_str(); }
    virtual ~exception() throw() { }
  protected:
    std::string message_;
  };
  */

  static const int FLOTSAM_BEAM_FLUX_TERMS = 5;
  static const int FLOTSAM_DIRECT = 0;
  static const int FLOTSAM_LOBE_DN_1 = 1;
  static const int FLOTSAM_LOBE_DN_2 = 2;
  static const int FLOTSAM_LOBE_UP_1 = 3;
  static const int FLOTSAM_LOBE_UP_2 = 4;

  static const adept::Real FLOTSAM_MAX_LAYER_OPTICAL_DEPTH = 20.0;

  // Fudge-factor accounting for enhanced scattering from diffuse
  // upward stream to lobe/direct returning beams
  static const adept::Real FLOTSAM_DIFFUSE_UP_FACTOR = 1.0;
  static const adept::Real FLOTSAM_DIFFUSE_DN_FACTOR = 1.0;

  // Preferred fudge factor
  //static const adept::Real FLOTSAM_NONDIRECT_SCALING = 1.05;
  static const adept::Real FLOTSAM_NONDIRECT_SCALING = 1.0;

  // Standard value for this was 0.5 until 0.5.13
  static const adept::Real FLOTSAM_LUT_OPTICAL_DEPTH_REF_FACTOR = 0.5;

  // Fraction of the forward lobe that is scattered into the
  // forward-hemisphere diffuse
  static const adept::Real FLOTSAM_FORWARD_LOBE_DIFFUSE = 0.0;
  //  static const adept::Real FLOTSAM_FORWARD_LOBE_DIFFUSE = 0.1;
  //  static const adept::Real FLOTSAM_FORWARD_LOBE_DIFFUSE = 0.2;
  //  static const adept::Real FLOTSAM_COS2_FORWARD_TO_LOBE = 0.5;
  static const adept::Real FLOTSAM_COS2_FORWARD_TO_LOBE = 0.5;
  //static const adept::Real FLOTSAM_COS2_FORWARD_TO_LOBE = 1.0;

  // Original value appropriate for pure exponential with angle 14deg
  //  static const adept::Real FLOTSAM_FORWARD_LOBE_REMAIN = 0.5846;
  //static const adept::Real FLOTSAM_FORWARD_LOBE_REMAIN = 0.2;
  static const adept::Real FLOTSAM_FORWARD_LOBE_REMAIN = 0.55;
  //  static const adept::Real FLOTSAM_FORWARD_LOBE_REMAIN = 0.4;
  // static const adept::Real FLOTSAM_FORWARD_LOBE_REMAIN = 0.8;
  // Value for exp-tan function with angle 18deg
  //  static const adept::Real FLOTSAM_FORWARD_LOBE_REMAIN = 0.734424657048408;
  //static const adept::Real FLOTSAM_FORWARD_LOBE_REMAIN = 0.0;

  //static const adept::Real FLOTSAM_MU_LOBE_UP  = 0.4;
  //  static const adept::Real FLOTSAM_MIN_DIFFUSIVITY_MU_LOBE_DN = 0.15;
  static const adept::Real FLOTSAM_MU_LOBE_UP  = 0.35;
  static const adept::Real FLOTSAM_MIN_DIFFUSIVITY_MU_LOBE_DN = 0.1;

  // Lobe angular width is predicted and used to select the level of
  // smoothing to use in the version of the phase function for the
  // lobe contribution
  static const int FLOTSAM_NUM_SMOOTHING_VALUES = 5;

  // Number of polynomial coefficients to use to describe the
  // smoothing dependence
  static const int FLOTSAM_NUM_SMOOTHING_COEFFS = 4;


}

#endif
